[nfc][llvm-profdata] Use cl::Subcommand to organize subcommand and options in llvm-profdata

mingmingl-llvm opened the pull request:
- The motivation is to reduce the number of arguments passed around (e.g., from `show_main` to `show*Profile`).  In order to do this, move function-defined options to global variables, and create `cl::SubCommand` for {show, merge, overlap, order} to organize options.
   - The side-effect by extracting function local options to a C++ namespace is that the extracted options are no longer (lazily) initialized when the enclosing function runs for the first time. 
   - `cl::Subcommand` support (introduced in https://lists.llvm.org/pipermail/llvm-dev/2016-June/101804.html) could put options in a per-subcommand namespace.
- One option could belong to multiple subcommand. This patch defines most of the options once and associates them with multiple subcommands except
   1. `overlap` and `show` both has `value-cutoff` with different default values ([former](https://github.com/llvm/llvm-project/blob/64f62de96609dc3ea9a8a914a9e9445b7f4d625d/llvm/tools/llvm-profdata/llvm-profdata.cpp#L2352) vs [latter](https://github.com/llvm/llvm-project/blob/64f62de96609dc3ea9a8a914a9e9445b7f4d625d/llvm/tools/llvm-profdata/llvm-profdata.cpp#L3009)). Define 'OverlapValueCutoff' and 'ShowValueCutoff' respectively.
   2. `show` supports three profile formats in `ProfileKind` while {`merge`, `overlap`} supports two. Define separate options.
- Clean up obsolete code as a result, including `-h` and `--version` customizations. These two options are supported for all commands.  Results pasted.
   - [-h and --help](https://gist.github.com/minglotus-6/387490e5eeda2dd2f9c440a424d6f360) output.
   - [--version](https://gist.github.com/minglotus-6/f905abcc3a346957bd797f2f84c18c1b)
   - [llvm-profdata show --help](https://gist.github.com/minglotus-6/f143079f02af243a94758138c0af471a)

This PR should be `llvm-profdata` only. It depends on https://github.com/llvm/llvm-project/pull/71981

mtrofin commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2406,6 +2407,14 @@ static void traverseAllValueSites(const InstrProfRecord &Func, uint32_t VK,
 
+namespace {
+struct InstrProfilePerFuncOptions {
+  bool ShowCounts;
```
please initialize scalars at declaration (i.e. `bool ShowCounts = false`), it's very easy to later write code that forgets to initialize a field, and while maybe noticing compiler warnings can highlight this, it's so much easier to just initialize at declaration and avoid undefined behavior!

minglotus-6 replied:
done.

mtrofin commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2420,14 +2429,97 @@ static void showValueSitesStats(raw_fd_ostream &OS, uint32_t VK,
+                       size_t &ShownFunctions, raw_fd_ostream &OS) {
+  if (!ShownFunctions)
+    OS << "Counters:\n";
+  ++ShownFunctions;
```
why not return this?

mtrofin replied:
(same below)

minglotus-6 replied:
It's currently an input and output parameter. I'm not a big fan (in terms of readability, etc) of returning an updated copy of const input parameter and assigning it to the same variable, compared with the current way of using input & output parameter.

mtrofin commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2420,14 +2429,97 @@ static void showValueSitesStats(raw_fd_ostream &OS, uint32_t VK,
+static void showValueProfileStats(size_t ShownFunctions,
+                                  bool ShowIndirectCallTargets,
+                                  bool ShowMemOPSizes,
+                                  std::vector<ValueSitesStats> &VPStats,
```
is VPStats not `const`-ed because of the use of `operator[]`? If so, is the desired behavior to insert a default value if `IPVK_MemOpSize` isn't a key?

minglotus-6 replied:
The vector 'VPStats' is initialized to as `std::vector<ValueSitesStats> VPStats(NumVPKind);` [code](https://github.com/llvm/llvm-project/blob/bc7a3bd864be696217c4d11eddf16bed7646b60f/llvm/tools/llvm-profdata/llvm-profdata.cpp#L2477), and out-of-bound access from `operator []` is [undefined behavior](https://cplusplus.com/reference/vector/vector/operator[]/).

I updated this new function and its callee to take a constant vector.

mtrofin commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2420,14 +2429,97 @@ static void showValueSitesStats(raw_fd_ostream &OS, uint32_t VK,
+}
+
+static int
+showInstrProfile(const std::string &Filename, uint32_t TopN,
```
I wonder if factoring all the `show...` APIs into a class would make more sense - do some of the parameters make sense as shared state between these functions?

minglotus-6 replied:
> I wonder if factoring all the show... APIs into a class would make more sense - do some of the parameters make sense as shared state between these functions?

To share some context of this NFC, the immediate (and motivating) use case `InstrProfilePerFuncOptions` is to allow showing one new type of value profiles without increasing the length of function arguments. 

`InstrProfileShowOptions` (a profile-level show options) would make sense as a separate patch. How to factor them is a more open-ended question. With all (nearly scalar) options, some options could be used together (e.g.,  `TopN` is meaningful to show function profiles but not meaningful in temporal profile traces)  while other options are more independent of each other.

david-xl commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2406,6 +2407,14 @@ static void traverseAllValueSites(const InstrProfRecord &Func, uint32_t VK,
 }
 
+namespace {
+struct InstrProfilePerFuncOptions {
```
Will it be limited to InstrProfile? Perhaps just name it "PerFuncOptions"?

minglotus-6 replied:
Showing indirect call targets seems implicit currently; `showSampleProfile` doesn't take an option to suppress indirect call targets. 

A minor thing is, functions in sample profiles doesn't have mem-op size; so passing 'PerFuncOptions' in `showSampleProfile` static function would require an option invalidation 

```
if (PerFuncOptions.ShowMemOPSizes)
  emit error with message "mem-op size is not supported in sample profiles"
```


The struct is in anonymous namespace which makes it easier for renames if necessary.

minglotus-6 commented:
The patch is currently adding new C++ structs to reduce the number of arguments.

Another option is to move option definitions (https://github.com/llvm/llvm-project/blob/45ca24edc0b9f2541efb0ae3c668ee7af82e22f4/llvm/tools/llvm-profdata/llvm-profdata.cpp#L2954-L3042) from function scope to a namespace (like `namespace show_prof{`), so the option definitions are visible to `showInstrProfile` and `showSampleProfile` (no need to pass long list of parameters) . The downside is that, not all options apply to both functions.

Thoughts about these two options?

david-xl commented:
> The patch is currently adding new C++ structs to reduce the number of arguments.
> 
> Another option is to move option definitions (
> 
> https://github.com/llvm/llvm-project/blob/45ca24edc0b9f2541efb0ae3c668ee7af82e22f4/llvm/tools/llvm-profdata/llvm-profdata.cpp#L2954-L3042
> 
> ) from function scope to a namespace (like `namespace show_prof{`), so the option definitions are visible to `showInstrProfile` and `showSampleProfile` (no need to pass long list of parameters) . The downside is that, not all options apply to both functions.
> Thoughts about these two options?

Moving show options to show_prof namespace looks reasonable to me. Need to make sure  the --help option for the 'show' command  works properly though.

mingmingl-llvm committed:
[nfc][llvm-profdata] Move show options to its namespace

minglotus-6 commented:
I updated the PR to exemplify the change for `show_main` and show options. After the refactor pattern is closer to what it should look like eventually, I could apply it to other static *main functions

david-xl reviewed (commented):
Better put the options into its own name space "namespace show_options {}. In the follow up patch, options for other commands can be done similarly.

minglotus-6 commented:
> Better put the options into its own name space "namespace show_options {}. In the follow up patch, options for other commands can be done similarly.

Sounds good. (Learnt from failed error message) I should probably register `show` of `llvm-profdata show` as a subcommand, and make the show options associated with this subcommand (something like this example https://github.com/llvm/llvm-project/blob/c2ad9f8b607931430e86da7420493599c48e62a0/llvm/tools/llvm-pdbutil/llvm-pdbutil.cpp#L163);

Without that, I'm getting runtime registration errors `Option 'output' registered more than once!` when extracting `output` option out of a per-function scope.

mingmingl-llvm committed:
Add show_options namespace and move show_options from function scope to
namespace.
- The change uses cl::SubCommand feature that allow to register options
  under a subcommand.
  - Without using cl::SubCommand, the options in the show name namespace will
    take a registered name as global variables, causing runtime errors when a
    function-scope static option tries to add a function of a same name.
  - Make changes in the CommandLine library accordingly.

minglotus-6 commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -3051,7 +3084,7 @@ static int show_main(int argc, const char *argv[]) {
   if (Filename == OutputFilename) {
-    errs() << sys::path::filename(argv[0])
+    errs() << sys::path::filename(argv[0]) << " " << argv[1]
            << ": Input file name cannot be the same as the output file name!\n";
```
this is just an artifact that `show` options parses command lines differently with the other three subcommands (that are not updated).

When all subcommands are updated to be `cl::SubCommand`, or when we introduce the following code, this is going to look more formal.

```
// If we want to update one subcommand at a time
if  subcommand uses cl::SubCommand {
   // the new method in this PR
   handles argv list one way
}
else {
  // the existing method
  handles argv list the other way
}
```

david-xl commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2448,101 +2449,14 @@ static void showValueSitesStats(raw_fd_ostream &OS, uint32_t VK,
-static int showInstrProfile(const std::string &Filename, ShowFormat SFormat,
-                            raw_fd_ostream &OS) {
+static int showInstrProfile(
+    const std::string &Filename, bool ShowCounts, uint32_t TopN,
```
Why still passing the options as parameters ? They can be referenced  as show_options::<OptionName> in the function body.

minglotus-6 replied:
I updated {showInstrProfile, showSampleProfile, showMemProfProfile, showDebugInfoCorrelation} to use options directly.

david-xl reviewed (commented):
Since the options are now attached to ShowSubCommand, there is no longer the need to use show_options namespace.

mingmingl-llvm committed:
resolve feedback

snehasish commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -1667,6 +1667,13 @@ bool CommandLineParser::ParseCommandLineOptions(int argc,
+      // If Handler is not found in a specialized subcommand, look up handler
+      // in the top-level subcommand.
+      // cl::opt without cl::sub belongs to top-level subcommand.
+      if (!Handler && ChosenSubCommand != &SubCommand::getTopLevel())
```
Lets split the changes llvm/Support/CommandLine out into a separate patch and add a test to cover this case. We can reference this PR as the motivation.

minglotus-6 replied:
> Lets split the changes llvm/Support/CommandLine out into a separate patch and add a test to cover this case. We can reference this PR as the motivation.

Sent out pr/71776 as a separate patch.

minglotus-6 commented:
> Since the options are now attached to ShowSubCommand, there is no longer the need to use show_options namespace.

Different subcommand could have `cl::opt` variables of the same name. While `cl::SubCommand` could put registered option strings in a sub-command "namespace", , the `cl::opt` variables are put in per-subcommand namespaces.

snehasish reviewed (commented):
Overall direction looks reasonable to me. Thanks for taking the time to improve the code here!

minglotus-6 commented:
> > The patch is currently adding new C++ structs to reduce the number of arguments.
> > Another option is to move option definitions (
> > https://github.com/llvm/llvm-project/blob/45ca24edc0b9f2541efb0ae3c668ee7af82e22f4/llvm/tools/llvm-profdata/llvm-profdata.cpp#L2954-L3042
> > 
> > ) from function scope to a namespace (like `namespace show_prof{`), so the option definitions are visible to `showInstrProfile` and `showSampleProfile` (no need to pass long list of parameters) . The downside is that, not all options apply to both functions.
> > Thoughts about these two options?
> 
> Moving show options to show_prof namespace looks reasonable to me. Need to make sure the --help option for the 'show' command works properly though.

With the usage of `cl::Subcommand`, `--help` should work out of the box.  This [gist](https://gist.github.com/minglotus-6/2be4603f0706a692dfe8990f57cf86dc) is the output.

> > Since the options are now attached to ShowSubCommand, there is no longer the need to use show_options namespace.
> 
> Different subcommand could have `cl::opt` variables of the same name. While `cl::SubCommand` could put registered option strings in a sub-command "namespace", , the `cl::opt` variables are put in per-subcommand namespaces.

I might need to take this back. It seems plausible to associate one `cl::opt` with multiple subcommands based on how registration works. I will try it out.

mingmingl-llvm committed:
Refactor llvm-profdata to use cl::subcommand

mingmingl-llvm committed:
undo one-line added during debugging

minglotus-6 commented:
This is ready for review again! 

Some options are sub-command specific and I tried to keep these options closer to where they are used. Common options are moved ahead at the very beginning.

minglotus-6 commented:
To ensure options are not deleted mistakenly, I run `llvm-profdata <subcommand> --help` with and without the patch, and got the diff links [1].  Two kinds of options are gone in emitted message with this patch, which I think is benign.
1) `--version` is not in `--help` message. note `llvm-profdata --version` still works but doesn't emit `llvm-profdata` anymore (see updated regression test)
2) Options that are not defined in `llvm-profdata.cpp` before this patch.  {disable-i2p-p2i-opt, disable-auto-upgrade-debug-info, enable-name-compression, generate-merged-base-profiles} are defined in dependent libraries.  `experimental-debuginfo-iterators` is shown before but I didn't see where this option is defined in the llvm-repo, so consider this fine.

[1] 
- order https://www.diffchecker.com/UokTAyYJ/
- overlap https://www.diffchecker.com/wO2rICsW/
- merge https://www.diffchecker.com/QyXCBpBO/
- show https://www.diffchecker.com/PKPMcPRB/

snehasish reviewed (approved):
lgtm with some minor comments.

minglotus-6 commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -435,13 +450,195 @@ static void writeInstrProfile(StringRef OutputFilename,
-                  uint64_t MaxTraceLength, int MaxDbgCorrelationWarnings,
-                  bool OutputSparse, unsigned NumThreads, FailureMode FailMode,
-                  const StringRef ProfiledBinary) {
+// Common options.
```
Sounds reasonable to me. I moved options definitions right after `cl::Subcommand`.

minglotus-6 commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -1362,28 +1413,55 @@ static int merge_main(int argc, const char *argv[]) {
+    cl::desc("For context sensitive PGO counts. Does not work with CSSPGO."),
+    cl::sub(OverlapSubcommand));
+
+cl::opt<std::string> FuncNameFilter(
```
good catch. I moved it to the 2-common options section.

mingmingl-llvm committed:
Merge branch 'main' into llvm-profdata

mingmingl-llvm committed:
resolve feedback

mingmingl-llvm reviewed (commented):
thanks for the reviews!

mingmingl-llvm committed:
move enum definition before option definition
